biodatasets/codiesp/codiesp.py

```diff
@@ -0,0 +1,310 @@
+        """
+        Downloads/extracts the data to generate the train, validation and test splits.
+
+        Each split is created by instantiating a `datasets.SplitGenerator`, which will
+        call `this._generate_examples` with the keyword arguments in `gen_kwargs`.
+        """
+
+        data_dir = dl_manager.download_and_extract(_URLS["codiesp"])
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "filepath": Path(os.path.join(data_dir, "final_dataset_v4_to_publish/train")),
+                    "split": "train",
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                gen_kwargs={
+                    "filepath": Path(os.path.join(data_dir, "final_dataset_v4_to_publish/test")),
+                    "split": "test",
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.VALIDATION,
+                gen_kwargs={
+                    "filepath": Path(os.path.join(data_dir, "final_dataset_v4_to_publish/dev")),
+                    "split": "dev",
+                },
+            ),
+        ]
+
+    def _generate_examples(self, filepath, split: str) -> Tuple[int, Dict]:
+        """
+        This method handles input defined in _split_generators to yield (key, example) tuples from the dataset.
+        Method parameters are unpacked from `gen_kwargs` as given in `_split_generators`.
+        """
+
+        paths = {"text_files": Path(os.path.join(filepath, "text_files"))}
+        for task in ["codiesp_d", "codiesp_p", "codiesp_x"]:
+            paths[task] = Path(os.path.join(filepath, f"{split}{task[-1].upper()}.tsv"))
+
+        if self.config.name == "codiesp_D_bigbio_text" or self.config.name == "codiesp_P_bigbio_text":
+            df = pd.read_csv(paths[self.config.subset_id], sep="\t", header=None)
+
+            file_codes_dict = defaultdict(list)
+            for idx, row in df.iterrows():
+                file, code = row[0], row[1]
+                file_codes_dict[file].append(code)
+
+            for guid, (file, codes) in enumerate(file_codes_dict.items()):
+                text_file = Path(os.path.join(paths["text_files"], f"{file}.txt"))
+                example = {
+                    "id": str(guid),
+                    "document_id": file,
+                    "text": text_file.read_text(),
+                    "labels": codes,
+                }
+                yield guid, example
+
+        elif self.config.name == "codiesp_X_bigbio_kb":
+            df = pd.read_csv(paths[self.config.subset_id], sep="\t", header=None)
+
+            example = {}
+            for guid, row in df.iterrows():
+                file, label, code, text, spans = row[0], row[1], row[2], row[3], row[4]
+                appearances = spans.split(";")
+                spans = []
```

Comment:
The way I understood it from [here (see CodiEsp Explainable AI exploratory) ](https://temu.bsc.es/codiesp/) is that the spans are "words in the text" that explain (are the reason why) a certain ICD10-CM/ICD10-PCS code was added.

To me this sounds like a `Tasks.NAMED_ENTITY_DISAMBIGUATION`. The dataset has lines in the form:
```bash
S0004-06142005000700014-1   PROCEDIMIENTO   bw03zzz Rx tórax   2163 2171
S0004-06142005000700014-1   DIAGNOSTICO n44.8   teste derecho aumentado de tamaño  1343 1376
```
So for instance I would convert the first line in `bigbio_kb` schema as:
```python
        "entities": [
            {
                "id": <unique id>,
                "type": "PROCEDIMIENTO", 
                "text": "Rx tórax",
                "offsets": [(2163,2171)],
                "normalized": [
                    {
                        "db_name": "ICD10-PCS",
                        "db_id": bw03zzz,
                    }
                ],
            }
        ],
```
Does it make sense?